gcloud-java-storage/src/main/java/com/google/gcloud/storage/Blob.java

```diff
@@ -0,0 +1,246 @@
+        convertedOptions[index++] = option.convert(blobInfo);
+      }
+      return convertedOptions;
+    }
+  }
+
+  /**
+   * Construct a {@code Blob} object for the provided {@code BlobInfo}. The storage service is used
+   * to issue requests.
+   * 
+   * @param storage the storage service used for issuing requests
+   * @param info blob's info
+   */
+  public Blob(Storage storage, BlobInfo info) {
+    this.storage = checkNotNull(storage);
+    this.info = checkNotNull(info);
+  }
+
+  /**
+   * Construct a {@code Blob} object for the provided bucket and blob names. The storage service is
+   * used to issue requests.
+   * 
+   * @param storage the storage service used for issuing requests
+   * @param bucket bucket's name
+   * @param blob blob's name
+   */
+  public Blob(Storage storage, String bucket, String blob) {
+    this.storage = checkNotNull(storage);
+    this.info = BlobInfo.of(bucket, blob);
+  }
+
+  /**
+   * Get the blobs's information.
+   * 
+   * @return a {@code BlobInfo} object for this blob
+   */
+  public BlobInfo info() {
+    return info;
+  }
+
+  /**
+   * Check if this blob exists.
+   *
+   * @return true if this blob exists, false otherwise
+   * @throws StorageException upon failure
+   */
+  public boolean exists() {
+    return storage.get(info.bucket(), info.name()) != null;
+  }
+
+  /**
+   * Return this blob's content.
+   *
+   * @param options blob read options
+   * @return the blob's content
+   * @throws StorageException upon failure
+   */
+  public byte[] content(Storage.BlobSourceOption... options) {
+    return storage.readAllBytes(info.bucket(), info.name(), options);
+  }
+
+  /**
+   * Update the blob's information. Bucket or blob's name cannot be changed by this method. If you
+   * want to rename the blob or move it to a different bucket use the {@link #copyTo} and
+   * {@link #delete} operations.
+   *
+   * @param blobInfo new blob's information. Bucket and blob names must match the current ones
+   * @param options update options
+   * @throws StorageException upon failure
+   */
+  public void update(BlobInfo blobInfo, BlobTargetOption... options) {
+    checkArgument(Objects.equals(blobInfo.bucket(), info.bucket()), "Bucket name must match");
+    checkArgument(Objects.equals(blobInfo.name(), info.name()), "Blob name must match");
+    info = storage.update(blobInfo, options);
+  }
+
+  /**
+   * Delete this blob.
+   *
+   * @param options blob delete options
+   * @return true if blob was deleted
+   * @throws StorageException upon failure
+   */
+  public boolean delete(BlobSourceOption... options) {
+    return storage.delete(info.bucket(), info.name(), convert(info, options));
+  }
+
+  /**
+   * Copy this blob.
+   *
+   * @param target target blob
+   * @param options source blob options
+   * @return the copied blob
+   * @throws StorageException upon failure
+   */
+  public Blob copyTo(BlobInfo target, BlobSourceOption... options) {
```

Comment:
Sorry for the tons of typos. For **gcloud-go** I meant that they have something similar to our `Storage.copy`.  
> Do they copy their own (this) metadata to the destination?

**gcloud-node**, **gcloud-ruby** and **gcloud-python** seem not to pass any metadata. **gcloud-go** has a metadata parameter that if specified is used to override metadata.

> But it does, using the values from target BlobInfo, just not from the source.

You are right.

> maybe we can start with something simpler such as.
copyTo(bucket, sourceOptions...) and copyTo(bucket, name, sourceOptions....)
and copy the metadata by default?

Absolutely. If we get rid of the `target` Blob parameter we can use current `Blob`'s info to possibly update metadata. Can we go with this?

> I consider the options as an advanced feature that most user will not care about and
would rather make it as transparent as possible. 

Agree. But that method is only there to allow a user to set both options, which as you said is not the common case. Normal users have their `copy(target)` right?

> Also, a side note, I am not sure the param sequence feels right (target, sourceOptions, targetOptions)..

Not sure any other permutation would make more sense :)